README.md
# Bedrock OAuth2 Token Request Handler _(@digitalbazaar/oauth2-token-endpoint-handler)_

[![Node.js CI](https://github.com/digitalbazaar/oauth2-token-endpoint-handler/workflows/Node.js%20CI/badge.svg)](https://github.com/digitalbazaar/oauth2-token-endpoint-handler/actions?query=workflow%3A%22Node.js+CI%22)

> OAuth2 Token endpoint handler for custom authorization servers.

## Table of Contents

- [Background](#background)
- [Security](#security)
- [Install](#install)
- [Usage](#usage)
- [Contribute](#contribute)
- [Commercial Support](#commercial-support)
- [License](#license)

## Background

Limitations / design constraints:

* Only supports the `client_credentials` grant.
* Only supports `client_secret_post` endpoint authentication method.
* Does not support Refresh Tokens (clients can just perform the token exchange
  again, when their original Access Token expires).

Inspired by:

 * [`oauth2orize`](https://github.com/jaredhanson/oauth2orize) and
   [`@passport-next/oauth2orize`](https://github.com/passport-next/oauth2orize)
 * [`oauth2orize-jwt-bearer`](https://github.com/xtuple/oauth2orize-jwt-bearer)
 * [`node-oidc-provider`](https://github.com/panva/node-oidc-provider)

Relevant specifications:

* [OAuth 2.0 Authorization Framework](https://tools.ietf.org/html/rfc6749)
* [Resource Indicators for OAuth 2.0](https://tools.ietf.org/html/rfc8707)

## Security

TBD

## Install

- Node.js >= 16 is required.

To install locally (for development):

```
git clone https://github.com/digitalbazaar/oauth2-token-endpoint-handler.git
cd oauth2-token-endpoint-handler
npm install
```

## Usage

```js
import {tokenExchangeHandler} from '@digitalbazaar/oauth2-token-endpoint-handler';

app.post('/token',
  // ... make sure to mount url-encoded body parser such as express.urlencoded()
  tokenExchangeHandler({
    defaults: {
      // (Optional) token expiration in seconds. Defaults to 2 weeks
      defaultMaxAge: 1209600
    },
    logger: console,
    getClient: async ({clientId}) => {
      // callback to load client from the db, then:
      return {client};
    },
    authenticateClient: async ({client, clientSecret}) => {
      // callback to authenticate client (compare hash of clientSecret to
      // what's stored, etc).
    },
    issue: async ({client, resource, scopeRequested, body, defaultMaxAge}) => {
      // callback to issue JWT access token
      return {accessToken, expiresIn};
    }
  })
);
```

Example request:

```
POST /token HTTP/1.1
Host: as.example.com
Content-Type: application/x-www-form-urlencoded

client_id=s6BhdRkqt3
&client_secret=7Fjfp0ZBr1KtDRbnfVdmIw
&grant_type=client_credentials
&scope=some.resource.read
&resource=https%3A%2F%2Fcalendar.example.com%2F
&resource=https%3A%2F%2Fcontacts.example.com%2F
```

Example response:

```
HTTP/1.1 200 OK
Content-Type: application/json;charset=UTF-8
Cache-Control: no-store
Pragma: no-cache

{
  "access_token": "eyJhbGciOiJFUzI1NiIsImtpZCI6Ijc3In0.eyJpc3MiOiJodHRwOi8vYXV0aG9yaXphdGlvbi1zZXJ2ZXIuZXhhbXBsZS5jb20iLCJzdWIiOiJfX2JfYyIsImV4cCI6MTU4ODQyMDgwMCwic2NvcGUiOiJjYWxlbmRhciIsImF1ZCI6Imh0dHBzOi8vY2FsLmV4YW1wbGUuY29tLyJ9.nNWJ2dXSxaDRdMUKlzs-cYIj8MDoM6Gy7pf_sKrLGsAFf1C2bDhB60DQfW1DZL5npdko1_Mmk5sUfzkiQNVpYw",
  "token_type": "Bearer",
  "expires_in": 3600,
  "scope": "some.resource.read"
}
```

## Contribute

See [the contribute file](https://github.com/digitalbazaar/bedrock/blob/master/CONTRIBUTING.md)!

PRs accepted.

If editing the Readme, please conform to the
[standard-readme](https://github.com/RichardLitt/standard-readme) specification.

## Commercial Support

Commercial support for this library is available upon request from
Digital Bazaar: support@digitalbazaar.com

## License

[New BSD License (3-clause)](LICENSE) © Digital Bazaar


CHANGELOG.md
# @digitalbazaar/oauth2-token-endpoint-handler ChangeLog

## 3.0.0 - 2023-02-06

### Changed
- **BREAKING**: Convert to module (ESM).
- **BREAKING**: Drop support for node <= 14.
- Update dev deps to latest.
- Use `c8@7.12.0` for coverage.

## 2.0.0 - 2021-12-15

### Added
- **BREAKING**: Added new parameter `cache` to `tokenExchangeHandler`. This is
  for security, as it limits the number of access tokens issued to each client.
  There is no forced cache clearing, so if a client secret is updated, the old
  access token will continue to be handed out until the cache releases it.

## 1.0.2 - 2020-10-11

### Changed
- Fix: Allow issuer to control `scope` of token (based on `scopeRequested`).

## 1.0.1 - 2020-10-11

### Changed
- Fix: Make `scope` parameter optional for `client_credentials` grant,
  per the OAuth 2.0 spec.

## 1.0.0 - 2020-10-09

- Initial release. See git history for changes.


tests/index.spec.js
/*!
 * Copyright (c) 2020-2023 Digital Bazaar, Inc. All rights reserved.
 */
import {_respond, tokenExchangeHandler} from '../lib/index.js';
import chai from 'chai';
import chaiHttp from 'chai-http';
import express from 'express';
import {InvalidClient} from '@interop/oauth2-errors';
import {LruCache} from '@digitalbazaar/lru-memoize';
import noopLogger from '../lib/noopLogger.js';

const cache = new LruCache({
  max: 10,
  // 5 minutes
  maxAge: 300000
});

chai.use(chaiHttp);
chai.should();

const {expect} = chai;
const bodyParserUrl = express.urlencoded({extended: true});

const tokenUrl = '/token';
const MOCK_ACCESS_TOKEN = 'eyJhbGciOiJFUzI1NiIsImtpZCI6Ijc3In0' +
  '.eyJpc3MiOiJodHRwOi8vYXV0aG9yaXphdGlvbi1zZXJ2ZXIuZXhhbXBsZS5jb20iLCJzdW' +
  'IiOiJfX2JfYyIsImV4cCI6MTU4ODQyMDgwMCwic2NvcGUiOiJjYWxlbmRhciIsImF1ZCI6I' +
  'mh0dHBzOi8vY2FsLmV4YW1wbGUuY29tLyJ9.nNWJ2dXSxaDRdMUKlzs-cYI' +
  'j8MDoM6Gy7pf_sKrLGsAFf1C2bDhB60DQfW1DZL5npdko1_Mmk5sUfzkiQNVpYw';

const mockAuthenticateClient = async ({client, clientSecret}) => {
  if(!client) {
    throw new InvalidClient({
      description: 'Unknown client identifier.'
    });
  }
  const {client_secret: storedClientSecret} = client;

  // NOTE: a timing safe comparison is required, this example is NOT secure
  return {authenticated: clientSecret === storedClientSecret};
};

const mockGetClient = async ({clientId = 's6BhdRkqt3'}) => {
  return {
    client_id: clientId,
    client_secret: '7Fjfp0ZBr1KtDRbnfVdmIw',
    scope: ''
  };
};

// eslint-disable-next-line no-unused-vars
const mockIssue = async ({client, resource, scope, body, defaultMaxAge}) => {
  return {
    accessToken: MOCK_ACCESS_TOKEN,
    expiresIn: 3600
  };
};

describe('tokenExchangeHandler', () => {
  const app = express();
  app.post(
    tokenUrl,
    bodyParserUrl,
    tokenExchangeHandler({
      authenticateClient: mockAuthenticateClient,
      getClient: mockGetClient,
      issue: mockIssue,
      cache
    })
  );

  let requester;

  before(async () => {
    requester = chai.request(app).keepOpen();
  });

  after(async () => {
    requester.close();
  });

  it(`should properly export noopLogger`, async () => {
    (typeof noopLogger === 'object').should.be.true;
  });
  it(`should properly export functions from main.js`, async () => {
    (typeof tokenExchangeHandler === 'function').should.be.true;
    (typeof _respond === 'function').should.be.true;
  });
  it('should error if missing client credentials', async () => {
    const res = await requester.post(tokenUrl).send({});
    expect(res).to.have.status(400);
    expect(res).to.be.json;
    expect(res.body.error).to.equal('unauthorized_client');
    expect(res.body.error_description)
      .to.equal('Missing client credentials at the Token endpoint.');
  });
  it('should error if missing grant_type param', async () => {
    const res = await requester.post(tokenUrl)
      .set('content-type', 'application/x-www-form-urlencoded')
      .send({
        client_id: 'some_id',
        client_secret: 'test_secret'
      });
    expect(res).to.have.status(400);
    expect(res).to.be.json;
    expect(res.body.error).to.equal('invalid_request');
    expect(res.body.error_description)
      .to.equal('Missing grant_type parameter.');
  });
  it('should error if grant_type is invalid', async () => {
    const res = await requester.post(tokenUrl)
      .set('content-type', 'application/x-www-form-urlencoded')
      .send({
        client_id: 'some_id',
        client_secret: 'test_secret',
        grant_type: 'invalid_type'
      });
    expect(res).to.have.status(400);
    expect(res).to.be.json;
    expect(res.body.error).to.equal('invalid_grant');
    expect(res.body.error_description)
      .to.equal('Grant type "invalid_type" is not supported.');
  });
  it('should error if grant_type is client_credentials and ' +
  'redirectUri is given', async () => {
    const res = await requester.post(tokenUrl)
      .set('content-type', 'application/x-www-form-urlencoded')
      .send({
        client_id: 'some_id',
        client_secret: 'test_secret',
        grant_type: 'client_credentials',
        redirect_uri: 'http://example.com'
      });
    expect(res).to.have.status(400);
    expect(res).to.be.json;
    expect(res.body.error).to.equal('invalid_grant');
    expect(res.body.error_description)
      .to.equal(
        'Grant type "client_credentials" cannot have "redirectUri" param.');
  });
  it('should error if client secret is invalid', async () => {
    const res = await requester.post(tokenUrl)
      .set('content-type', 'application/x-www-form-urlencoded')
      .send({
        client_id: 'some_id',
        client_secret: 'invalid_secret',
        grant_type: 'client_credentials',
      });
    expect(res).to.have.status(401);
    expect(res).to.be.json;
    expect(res.body.error).to.equal('invalid_client');
    expect(res.body.error_description).to.equal(
      'The client secret could not be authenticated.');
  });
  it('should return a valid access token with correct client credentials',
    async () => {
      const res = await requester.post(tokenUrl)
        .set('content-type', 'application/x-www-form-urlencoded')
        .send({
          client_id: 'some_id',
          client_secret: '7Fjfp0ZBr1KtDRbnfVdmIw',
          grant_type: 'client_credentials',
        });
      expect(res).to.have.status(201);
      expect(res).to.be.json;
      expect(res.body.access_token).to.be.a('string');
      expect(res.body.token_type).to.be.a('string');
      expect(res.body.token_type).to.equal('Bearer');
      expect(res.body.expires_in).to.be.a('number');
      expect(res.body.expires_in).to.equal(3600);
    });
  it('should error if it fails to load client', async () => {
    const badApp = express();
    badApp.post(
      tokenUrl,
      bodyParserUrl,
      tokenExchangeHandler({
        authenticateClient: mockAuthenticateClient,
        getClient: () => {
          throw new Error('Failed to get client.');
        },
        issue: mockIssue
      })
    );
    const res = await chai.request(badApp).post(tokenUrl)
      .set('content-type', 'application/x-www-form-urlencoded')
      .send({
        client_id: 'some_id',
        client_secret: '7Fjfp0ZBr1KtDRbnfVdmIw',
        grant_type: 'client_credentials',
      });
    expect(res).to.have.status(400);
    expect(res).to.be.json;
    expect(res.body.error).to.equal('invalid_request');
    expect(res.body.error_description).to.equal('Could not load client.');
  });
});


lib/main.js
/*!
 * Copyright (c) 2020 Digital Bazaar, Inc. All rights reserved.
 */
import assert from 'assert-plus';
import noopLogger from './noopLogger';
import {
  UnauthorizedClient, InvalidGrant, InvalidRequest, InvalidClient
} from '@interop/oauth2-errors';

// Default Access token expiration in seconds (two weeks)
const DEFAULT_MAX_AGE = 1209600;

/**
 * Constructs a middleware handler for OAuth 2.0 /token exchange requests.
 *
 * @param {object} options - Options hashmap.
 * @param {Function} options.authenticateClient - Async callback that
 *   authenticates the client token.
 * @param {object} [options.defaults={}] - Token issuance defaults.
 * @param {object} [options.logger] - Logger.
 * @param {object} options.cache - Cache for the access token. There is no
 *   forced cache clearing, so if a client secret is updated, the old
 *   access token will continue to be handed out until the cache releases it.
 * @param {Function} options.getClient - Async callback to load client by ID.
 * @param {Function} options.issue - Async callback that issues the JWT access
 *   token for the client.
 *
 * @example <caption>Example usage.</caption>
 * tokenExchangeHandler({
 *  defaults: {},
 *  logger: console,
 *  getClient: async ({clientId}) => {
 *    // load client from the db, then:
 *    return {client};
 *  },
 *  issue: async ({client, scope, body, defaultMaxAge}) => {
 *    // issue JWT access token here
 *    return {accessToken, expiresIn};
 *  },
 * });
 *
 * @returns {Function<Promise>} Resolves with exchange middleware.
 */
function tokenExchangeHandler({
  authenticateClient,
  defaults = {},
  getClient,
  issue,
  cache,
  logger = noopLogger,
}) {
  assert.func(authenticateClient, 'authenticateClient');
  assert.func(getClient, 'getClient');
  assert.func(issue, 'issue');

  return async (req, res) => {
    try {
      const {body} = req;
      const {
        client_id: clientId,
        client_secret: clientSecret,
        grant_type: grantType,
        scope: scopeRequested,
        redirect_uri: redirectUri
      } = body;

      _validateParams({clientId, clientSecret, grantType, redirectUri});

      let client;
      try {
        client = await getClient({clientId});
      } catch(cause) {
        const error = new InvalidRequest({
          description: 'Could not load client.'
        });
        error.cause = cause;
        throw error;
      }

      const {authenticated} = await authenticateClient({client, clientSecret});
      if(!authenticated) {
        throw new InvalidClient({
          description: 'The client secret could not be authenticated.'
        });
      }

      const defaultMaxAge = defaults.defaultMaxAge || DEFAULT_MAX_AGE;

      const key = JSON.stringify([client, scopeRequested, body, defaultMaxAge]);
      const fn = () => issue({
        client, scopeRequested, body, defaultMaxAge
      });

      const {accessToken, expiresIn, scope} = await cache.memoize({key, fn});

      const response = {
        access_token: accessToken,
        token_type: 'Bearer',
        expires_in: expiresIn,
        scope
      };

      _respond({response, res});
    } catch(error) {
      _error({error, logger, res});
    }
  };
}

function _validateParams({clientId, clientSecret, grantType, redirectUri}) {
  if(!(clientId && clientSecret)) {
    throw new UnauthorizedClient({
      description: 'Missing client credentials at the Token endpoint.'
    });
  }
  if(!grantType) {
    throw new InvalidRequest({
      description: 'Missing grant_type parameter.'
    });
  }
  if(grantType !== 'client_credentials') {
    throw new InvalidGrant({
      error: 'unsupported_grant_type',
      description: `Grant type "${grantType}" is not supported.`
    });
  }
  if(grantType === 'client_credentials' && redirectUri) {
    throw new InvalidGrant({
      error: 'unsupported_client_credentials_grant',
      description: 'Grant type "client_credentials" cannot ' +
        'have "redirectUri" param.'
    });
  }
}

function _respond({response, res}) {
  res.set({
    'Cache-Control': 'no-store',
    Pragma: 'no-cache'
  });

  res.status(201).json(response);
}

function _error({error, logger, res}) {
  logger.error('Token exchange endpoint error.', {error});
  const {
    error: errorId, error_description: description, error_uri: uri
  } = error;

  const statusCode = error.statusCode || 400;
  const oauth2ErrorResponse = {
    error: errorId || 'invalid_request',
    error_description: description || error.message,
    error_uri: uri
  };
  res.set({
    'Cache-Control': 'no-store',
    Pragma: 'no-cache'
  });
  return res.status(statusCode).json(oauth2ErrorResponse);
}

export {
  tokenExchangeHandler,
  _respond
};


lib/index.js
/*!
 * Copyright (c) 2020-2023 Digital Bazaar, Inc. All rights reserved.
 */
import {
  InvalidClient, InvalidGrant, InvalidRequest, UnauthorizedClient
} from '@interop/oauth2-errors';
import assert from 'assert-plus';
import noopLogger from './noopLogger.js';

// Default Access token expiration in seconds (two weeks)
const DEFAULT_MAX_AGE = 1209600;

/**
 * Constructs a middleware handler for OAuth 2.0 /token exchange requests.
 *
 * @param {object} options - Options hashmap.
 * @param {Function} options.authenticateClient - Async callback that
 *   authenticates the client token.
 * @param {object} [options.defaults={}] - Token issuance defaults.
 * @param {object} [options.logger] - Logger.
 * @param {object} options.cache - Cache for the access token. There is no
 *   forced cache clearing, so if a client secret is updated, the old
 *   access token will continue to be handed out until the cache releases it.
 * @param {Function} options.getClient - Async callback to load client by ID.
 * @param {Function} options.issue - Async callback that issues the JWT access
 *   token for the client.
 *
 * @example <caption>Example usage.</caption>
 * tokenExchangeHandler({
 *  defaults: {},
 *  logger: console,
 *  getClient: async ({clientId}) => {
 *    // load client from the db, then:
 *    return {client};
 *  },
 *  issue: async ({client, scope, body, defaultMaxAge}) => {
 *    // issue JWT access token here
 *    return {accessToken, expiresIn};
 *  },
 * });
 *
 * @returns {Function<Promise>} Resolves with exchange middleware.
 */
function tokenExchangeHandler({
  authenticateClient,
  defaults = {},
  getClient,
  issue,
  cache,
  logger = noopLogger,
}) {
  assert.func(authenticateClient, 'authenticateClient');
  assert.func(getClient, 'getClient');
  assert.func(issue, 'issue');

  return async (req, res) => {
    try {
      const {body} = req;
      const {
        client_id: clientId,
        client_secret: clientSecret,
        grant_type: grantType,
        scope: scopeRequested,
        redirect_uri: redirectUri
      } = body;

      _validateParams({clientId, clientSecret, grantType, redirectUri});

      let client;
      try {
        client = await getClient({clientId});
      } catch(cause) {
        const error = new InvalidRequest({
          description: 'Could not load client.'
        });
        error.cause = cause;
        throw error;
      }

      const {authenticated} = await authenticateClient({client, clientSecret});
      if(!authenticated) {
        throw new InvalidClient({
          description: 'The client secret could not be authenticated.'
        });
      }

      const defaultMaxAge = defaults.defaultMaxAge || DEFAULT_MAX_AGE;

      const key = JSON.stringify([client, scopeRequested, body, defaultMaxAge]);
      const fn = () => issue({
        client, scopeRequested, body, defaultMaxAge
      });

      const {accessToken, expiresIn, scope} = await cache.memoize({key, fn});

      const response = {
        access_token: accessToken,
        token_type: 'Bearer',
        expires_in: expiresIn,
        scope
      };

      _respond({response, res});
    } catch(error) {
      _error({error, logger, res});
    }
  };
}

function _validateParams({clientId, clientSecret, grantType, redirectUri}) {
  if(!(clientId && clientSecret)) {
    throw new UnauthorizedClient({
      description: 'Missing client credentials at the Token endpoint.'
    });
  }
  if(!grantType) {
    throw new InvalidRequest({
      description: 'Missing grant_type parameter.'
    });
  }
  if(grantType !== 'client_credentials') {
    throw new InvalidGrant({
      error: 'unsupported_grant_type',
      description: `Grant type "${grantType}" is not supported.`
    });
  }
  if(grantType === 'client_credentials' && redirectUri) {
    throw new InvalidGrant({
      error: 'unsupported_client_credentials_grant',
      description: 'Grant type "client_credentials" cannot ' +
        'have "redirectUri" param.'
    });
  }
}

function _respond({response, res}) {
  res.set({
    'Cache-Control': 'no-store',
    Pragma: 'no-cache'
  });

  res.status(201).json(response);
}

function _error({error, logger, res}) {
  logger.error('Token exchange endpoint error.', {error});
  const {
    error: errorId, error_description: description, error_uri: uri
  } = error;

  const statusCode = error.statusCode || 400;
  const oauth2ErrorResponse = {
    error: errorId || 'invalid_request',
    error_description: description || error.message,
    error_uri: uri
  };
  res.set({
    'Cache-Control': 'no-store',
    Pragma: 'no-cache'
  });
  return res.status(statusCode).json(oauth2ErrorResponse);
}

export {
  tokenExchangeHandler,
  _respond
};
